mingpt/trainer.py

```diff
@@ -56,11 +57,37 @@ def save_checkpoint(self):
             logger.info("saving %s", self.config.ckpt_path)
             torch.save(ckpt_model.state_dict(), self.config.ckpt_path)
 
+    def parameter_names_for_module_types(self, model, module_types):
+        parameters = set()
+        for n, m in model.named_modules():
+            for module_type in module_types:
+                if isinstance(m, module_type):
+                    for pn, p in m.named_parameters():
+                        parameters.add(n + "." + pn)
+        return list(parameters)
+
+    def no_decay_parameter_names(self, model, decay_parameter_name_overrides=None):
+
+        # define module types we always want to exclude from weight decay
+        no_decay_module_types = (torch.nn.modules.normalization.LayerNorm, torch.nn.modules.sparse.Embedding)
+
+        all_module_parameter_names = self.parameter_names_for_module_types(model, (torch.nn.Module,))
+
+        freestanding_parameter_names = [n for n, p in model.named_parameters() if not any(nd in n for nd in all_module_parameter_names)]
```

Comment:
I deliberately didn't use named_parameters(recurse=False) here because we want to find *all* freestanding parameter_names,  and not just those defined at the top-level.